web_poet/rules.py

```diff
@@ -0,0 +1,352 @@
+from __future__ import annotations  # https://www.python.org/dev/peps/pep-0563/
+
+import importlib
+import importlib.util
+import pkgutil
+import warnings
+from collections import deque
+from operator import attrgetter
+from typing import Any, Dict, Iterable, List, Optional, Type, TypeVar, Union
+
+import attrs
+from url_matcher import Patterns
+
+from web_poet._typing import get_item_cls
+from web_poet.pages import ItemPage
+from web_poet.utils import _create_deprecated_class, as_list, str_to_pattern
+
+Strings = Union[str, Iterable[str]]
+
+PageObjectRegistryTV = TypeVar("PageObjectRegistryTV", bound="PageObjectRegistry")
+
+
+@attrs.define(frozen=True)
+class ApplyRule:
+    """A rule that primarily applies Page Object and Item overrides for a given
+    URL pattern.
+
+    This is instantiated when using the :func:`web_poet.handle_urls` decorator.
+    It's also being returned as a ``List[ApplyRule]`` when calling the
+    ``web_poet.default_registry``'s :meth:`~.PageObjectRegistry.get_rules`
+    method.
+
+    You can access any of its attributes:
+
+        * ``for_patterns`` - contains the list of URL patterns associated with
+          this rule. You can read the API documentation of the `url-matcher
+          <https://url-matcher.readthedocs.io/>`_ package for more information
+          about the patterns.
+        * ``use`` - The Page Object that will be **used** in cases where the URL
+          pattern represented by the ``for_patterns`` attribute is matched.
+        * ``instead_of`` - *(optional)* The Page Object that will be **replaced**
+          with the Page Object specified via the ``use`` parameter.
+        * ``to_return`` - *(optional)* The Item Class that marks the Page Object
+          to be **used** which is capable of returning that Item Class.
+        * ``meta`` - *(optional)* Any other information you may want to store.
+          This doesn't do anything for now but may be useful for future API updates.
+
+    The main functionality of this class lies in the ``instead_of`` and ``to_return``
+    parameters. Should both of these be omitted, then :class:`~.ApplyRule` simply
+    tags which URL patterns the given Page Object is expected to be used.
+
+    As much as possible, the ``to_return`` parameter should capture the Item Class
+    that the Page Object is capable of returning. Before passing it to :class:`~.ApplyRule`,
+    the ``to_return`` value is primarily derived from the return class specified
+    from Page Objects that are subclasses of :class:`~.ItemPage`. However, a
+    special case exists when a Page Object returns a ``dict`` as an item but then
+    the rule should have ``to_return=None`` and **NOT** ``to_return=dict``.
```

Comment:
Some questions that I still have about `to_return` after reading the whole docstring:

- Is `to_return` a replacement for `use`? Does it make sense to use both parameters? When to use one an not the other?

> As much as possible, the ``to_return`` parameter should capture the Item Class that the Page Object is capable of returning.

- Which page object, the one assigned to “use”? Why “as much as possible”, when would it make sense to set a different `to_return` value? Is this about the `dict` case mentioned later, or is there more to it?

> Before passing it to :class:`~.ApplyRule`

Should we replace “Before” with “When not”? Or does it change the meaning?

> primarily derived

Given “primarily” here, I assume there are other cases, and we should cover those here.

> the return class specified from Page Objects that are subclasses of :class:`~.ItemPage`

I think users reading this may not understand how those return classes are specified. Maybe we could link “specified” to a documentation piece covering how to declare the return type of a Page Object by subclassing ItemPage.

Reply:
Thanks for raising these! I've improved the docs to help alleviate the confusion and ambiguity on them. Let me know what you think! https://github.com/scrapinghub/web-poet/pull/84/commits/776cf0dd16a466504e64875307da86d4ee2a6115

> Which page object, the one assigned to “use”? Why “as much as possible”, when would it make sense to set a different to_return value? Is this about the dict case mentioned later, or is there more to it?

I removed the “as much as possible" phrase here to remove the ambiguity. We also want to encourage the use of declaring the item class via `ItemPage[Product]`.

> Should we replace “Before” with “When not”? Or does it change the meaning?

Let's stick with "Before" as it means that the value should be derived first before before instantiating `ApplyRule`.

> Given “primarily” here, I assume there are other cases, and we should cover those here.

The "primarily" suggests that we can't actually derive everything completely from the return class. For example, if the PO returns `dict`, we set it to `to_return=None`. I'm not sure if we should cover other advanced use cases apart from this special case as early as now. I'm also not sure what are the other special scenarios that deriving something won't actually work. We can cover that later on as we and our users stumble across them.

> I think users reading this may not understand how those return classes are specified. Maybe we could link “specified” to a documentation piece covering how to declare the return type of a Page Object by subclassing ItemPage.

Good idea. I've added a link pointing to the `Item Class` subsection in the Overrides tutorial.